### PR TITLE
[WIP] Using UniFFi bindgen as a cargo subcommand

### PR DESCRIPTION
**Pull request by skhamis**
Discussion #472 

Taking a crack at making a more robust UniFFi CLI that allows for more flexibility as well as potential new tooling for checking environment to ensure success when building using UniFFi.

Suggestions welcome!

**To install**: In the bindgen dir -> `cargo install --path .` will install two binaries `uniffi-bindgen` and `cargo-uniffi` 

**To use**: 
- (used today) `uniffi-bindgen <commands>` 
- (new subcommand) `cargo uniffi <commands>`

New args:
` cargo uniffi check -l [language]` will ensure that you have things properly installed
(in progress) ` cargo uniffi check ?` will check scaffolding and udl files are properly setup

Planned changes:

- [X] Ability to have uniffi-bindgen as a subcommand ex `cargo uniffi scaffolding <udl>`
- [X] Adding a check command to check if local environment can succeed in building the foreign language bindings
- [X] Adding checks to the build process to ensure the udl is correct
- [ ] Check to see if there are not conflicting uniffi-bindgens installed(?)
- [ ] Attempt to integrate with the work in #997 
  > The possibility of split back-ends will mean we potentially should look into allowing "plugin-like" format for different bindgen into the core and defer to those bindgens via [external subcommands](https://docs.rs/clap/2.31.1/clap/struct.ArgMatches.html#method.subcommand) to allow language-specific crates to have their own set of commands

### CODE REVIEW

File: uniffi_bindgen/src/lib.rs [hunk 1]
```diff
@@ -424,11 +358,125 @@ pub fn run_main() -> Result<()> {
...
+            println!("Are we even here!?!?");
+            crate::run_checks(m.values_of("language").unwrap().collect())?
+        }
         _ => bail!("No command specified; try `--help` for some help."),
     }
     Ok(())
 }
 
+fn gen_comands<'a>() -> ArgMatches<'a> {
+    const POSSIBLE_LANGUAGES: &[&str] = &["kotlin", "python", "swift", "gecko_js", "ruby"];
```
[the remark this answers is not in the document]
Comment (skhamis) [reply uniffi_bindgen/src/lib.rs:371]: hmmm yes that's definitely something that's been in the back of my mind during these meetings -- I think that might be the most straightforward way to implement it. One thought I had is the idea of "plugins" to the uniffi bindgen that allows the commands from other bindgens to be "imported" to the main bindgen. I got inspired from the bevy game engine:
https://bevyengine.org/learn/book/getting-started/plugins/

_Of course it wouldn't look like how their doing it but throwing some some wild ideas out there._

File: uniffi_bindgen/Cargo.toml [hunk 2]
```diff
@@ -11,7 +11,7 @@ edition = "2018"
 keywords = ["ffi", "bindgen"]
 
 [[bin]]
-name = "uniffi-bindgen"
```
[the remark this answers is not in the document]
Comment (skhamis) [reply uniffi_bindgen/Cargo.toml:14]: This is actually a little more tricky than appears on the surface. The three ways it seems we can get this done is:
(1) in the cargo.toml have
```
[[bin]]
name = "cargo-uniffi"
path = "src/main.rs"

[[bin]]
name = "uniffi-bindgen"
path = "src/main.rs"
```
which will produce an warning `warning: .../uniffi-rs/uniffi_bindgen/Cargo.toml: file found to be present in multiple build targets: .../uniffi-rs/uniffi_bindgen/src/main.rs`

which may not be an issue as we are intentionally doing this...

(2) Have a second main.rs in a separate directory that just calls the `mod uniffi_bindgen`

(3) Create a shared lib and then have two mains use the same shared library https://github.com/rust-lang/cargo/issues/5930

It seems 3 is the "recommended" solution but open to any ideas!

[the remark this answers is not in the document]
Comment (rfk) [reply uniffi_bindgen/Cargo.toml:14]: > (3) Create a shared lib and then have two mains use the same shared library rust-lang/cargo#5930

Yeah, I think if I were a new developer coming into the project, this is what I'd expect to see - a shared lib with the functionality and then several small "wrapper" crates that use it to implement a specific command-line tool.

**Comment by skhamis** [comment]
Trying to write some "production-ready" code and thinking about this just a tad bit more. I wonder if it makes sense instead of having a `cargo uniffi check` command with all different args underneath it. We actually include a different arg called `--dry-run` that runs within the generate, scaffolding, etc commands which will then check the environment and that everything compiles fine with helpful statements instead of a side check....thoughts @rfk (and of course anyone else)?

**Comment by rfk** [comment]
> We actually include a different arg called --dry-run that runs within the generate, scaffolding, etc commands
> which will then check the environment and that everything compiles fine with helpful statements instead of a side check

It sounds worth a try! One thing about a separate `check` command could be that it's in a position to do more expensive checks than you might want in your build process by default (for example, it could shell out to `cargo metadata` and look at all the crates in your dependency tree in order to check that things look OK, while you might prefer that your day-to-day build process avoid this overhead by assuming everything is properly configured). But, I guess the `--dry-run` flag could also enable such checks!

**Comment by rfk** [comment]
> Ability to have uniffi-bindgen as a subcommand ex cargo uniffi scaffolding <udl>

So here's an example of something that I think would be a really interesting direction for UniFFI - what if I didn't need to specify the `<udl>` path explicitly here? That is, what if `cargo uniffi scaffolding` did the right thing without having to specify any options?

There are a number of ways we could achieve that:

* Configure the path to the `.udl` file in `Cargo.toml` and/or `uniffi.toml`.
* Presume some default filename like `./src/interface.udl` if one isn't specified.
* (...eventually get rid of UDL files altogether in favour of macros 😁)

What I'm interested in is the big-picture shift from pointing at an interface file, to pointing at the crate itself.

**Review by mhammond — CHANGES_REQUESTED** [review]
Not sure if we will pick this back up or not, but requesting changes to not have it show up as awaiting review.

**Comment by jplatte** [comment]
Given the progressing work on #1257, it might not sure a whole lot of sense to continue work on improving UDL-related work flows, which are likely going to be replaced entirely by that in a few months.

**Comment by badboy** [comment]
the uniffi-bindgen binary is since gone, so this would need some rethinking. I think the best for now is to close it.

**Pull request closed without merging**